## 💻 환경 설정

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
import folium
import sys

import geopy
from tqdm import tqdm

from pycaret.regression import *

#경고창이 뜨지 않도록 해주는 것
import warnings
warnings.filterwarnings('ignore')

# notebook을 실행한 브라우저에서 바로 그림을 볼 수 있게 해주는 것
%matplotlib inline 

>주피터 노트북에서 시각화할때, matplotlib가 한글 폰트 지원하지 않아, 깨짐 처리해줘야 함

In [2]:
import platform                

from matplotlib import font_manager, rc
plt.rcParams['axes.unicode_minus']= False

if platform.system() == 'Darwin': #맥os 사용자의 경우에
    plt.style.use('seaborn-darkgrid') 
    rc('font', family = 'AppleGothic')
    
elif platform.system() == 'Windows':#윈도우 사용자의 경우에
    path = 'c:/Windows/Fonts/malgun.ttf'
    font_name = font_manager.FontProperties(fname=path).get_name()
    plt.style.use('seaborn-darkgrid') # https://python-graph-gallery.com/199-matplotlib-style-sheets/
    rc('font', family=font_name)

> pandas 작업 중 tqdm을 사용하기 위해 다음과 같은 설정을 합니다.

In [3]:
tqdm.pandas()

>복원성을 위해 아래 random_state를 상수로 고정합니다.

In [4]:
RANDOM_STATE= 0
np.random.seed(RANDOM_STATE)

##  💻데이터 불러오기 

In [5]:
df = pd.read_csv('./NYC_taxi') 

In [6]:
df.head()

zip_code          pickup_hour  month  day  weekday  hour  is_weekend  cnt
0     11106  2015-01-01 00:00:00      1    1        3     0           0   75
1     10006  2015-01-01 00:00:00      1    1        3     0           0   64
2     11201  2015-01-01 00:00:00      1    1        3     0           0  169
3     11430  2015-01-01 00:00:00      1    1        3     0           0  222
4     10032  2015-01-01 00:00:00      1    1        3     0           0   64

## 💻 데이터 파악 및 이상유무 확인

In [7]:
df.shape

(87020, 8)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87020 entries, 0 to 87019
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   zip_code     87020 non-null  int64 
 1   pickup_hour  87020 non-null  object
 2   month        87020 non-null  int64 
 3   day          87020 non-null  int64 
 4   weekday      87020 non-null  int64 
 5   hour         87020 non-null  int64 
 6   is_weekend   87020 non-null  int64 
 7   cnt          87020 non-null  int64 
dtypes: int64(7), object(1)
memory usage: 5.3+ MB


> 결측치가 하나도 없습니다.

In [9]:
df['zip_code'].describe()

count    87020.000000
mean     10567.760296
std        579.688987
min      10001.000000
25%      10032.000000
50%      10177.000000
75%      11217.000000
max      14801.000000
Name: zip_code, dtype: float64

In [10]:
df['zip_code'].value_counts()

10019    744
10028    744
10002    743
10036    743
10128    743
10065    743
10029    743
10021    743
10010    743
10022    743
11104    742
10003    742
11101    742
10023    742
10014    742
10013    742
10017    742
10025    741
10016    741
10018    741
10024    741
10011    741
10012    741
10075    741
10009    741
10035    740
10026    740
10001    740
10031    739
11377    739
11106    739
11102    739
10038    739
10280    738
10027    738
10007    738
11201    738
10004    737
11211    737
10020    737
11217    737
10119    737
11430    737
10032    736
10006    736
10005    736
11215    736
11103    735
11231    734
10199    733
10030    733
10170    733
10282    732
11222    730
10110    725
10174    724
11105    723
11205    722
10168    721
10165    719
10153    718
11238    717
10037    715
10103    714
10172    713
11372    713
10171    712
11373    709
11371    707
10069    705
10152    705
10167    697
11369    693
10154    693
10451    693
10033    693
11370    689

In [11]:
df['zip_code'].value_counts().describe()

count    374.000000
mean     232.673797
std      301.011504
min        1.000000
25%        4.000000
50%       29.000000
75%      588.000000
max      744.000000
Name: zip_code, dtype: float64

In [12]:
# plt.figure(figsize=(15,8))
# sns.histplot(x='zip_code', data=df, discrete=True)

> zip_code는 총 374 개의 값이 있으며, 분포에 꽤 차이가 있습니다.

In [13]:
df['zip_code'].value_counts().loc[df['zip_code'].value_counts() > 100]

10019    744
10028    744
10002    743
10036    743
10128    743
10065    743
10029    743
10021    743
10010    743
10022    743
11104    742
10003    742
11101    742
10023    742
10014    742
10013    742
10017    742
10025    741
10016    741
10018    741
10024    741
10011    741
10012    741
10075    741
10009    741
10035    740
10026    740
10001    740
10031    739
11377    739
11106    739
11102    739
10038    739
10280    738
10027    738
10007    738
11201    738
10004    737
11211    737
10020    737
11217    737
10119    737
11430    737
10032    736
10006    736
10005    736
11215    736
11103    735
11231    734
10199    733
10030    733
10170    733
10282    732
11222    730
10110    725
10174    724
11105    723
11205    722
10168    721
10165    719
10153    718
11238    717
10037    715
10103    714
10172    713
11372    713
10171    712
11373    709
11371    707
10069    705
10152    705
10167    697
11369    693
10154    693
10451    693
10033    693
11370    689

> 100번 이상 등장하는 값이 154개로, 절반 이상이 100회가 되지 않습니다.

## 💻 위치 정보 가져오기

In [ ]:
import geopy

In [14]:
geolocator = geopy.Nominatim(user_agent='gh-quest')
geolocator.geocode({"postalcode": 10001}, country_codes="us").raw['lat']
geolocator.geocode({"postalcode": 10001}, country_codes="us").raw['lon']

'-73.99411702534721'

In [15]:
def postcode_to_lat(row):
    postcode = row['zip_code']
    try:
        lat = geolocator.geocode({"postalcode": postcode}, country_codes="us").raw['lat']
    except:
        lat = 0
    return lat

def postcode_to_lon(row):
    postcode = row['zip_code']
    try:
        lon = geolocator.geocode({"postalcode": postcode}, country_codes="us").raw['lon']
    except:
        lon = 0
    return lon

In [20]:
zip_unique = pd.DataFrame(df['zip_code'].unique(), columns=['zip_code'])
zip_unique

zip_code
0       11106
1       10006
2       11201
3       11430
4       10032
5       10075
6       10069
7       10001
8       10019
9       11216
10      10169
11      10455
12      10025
13      11218
14      10016
15      11205
16      10162
17      11103
18      10022
19      10128
20      10020
21      10010
22      11104
23      10030
24      11207
25      11419
26      10452
27      10472
28      10279
29      11208
30      11420
31      11361
32      11434
33      11212
34      10031
35      11206
36      10111
37      10005
38      10103
39      10028
40      10278
41      11385
42      10012
43      11416
44      11356
45      10282
46      11435
47      11424
48      10461
49      11235
50      11236
51      10458
52      11373
53      10065
54      10014
55      10024
56      11237
57      11109
58      10110
59      11231
60      11377
61      11213
62      11211
63      10168
64      10034
65      11367
66      10011
67      10153
68      10171
69      10454
70      10119
71      10165
72      10167
73      11233
74      11225
75      10029
76      11217
77      11514
78      10036
79      11368
80      11370
81      11372
82      10468
83      10154
84      11374
85      10453
86      11378
87      11415
88      10462
89      11230
90      11210
91      11433
92      11101
93      10035
94      10280
95      11749
96      10023
97      10033
98      10013
99      10017
100     10463
101     10007
102     11788
103     10173
104     10456
105     10457
106     10040
107     11369
108     10044
109     10473
110     11209
111     11226
112     11379
113     11220
114     11050
115     11215
116     10038
117     10018
118     10451
119     10026
120     10027
121     10172
122     11222
123     10199
124     10021
125     10177
126     11354
127     10152
128     10469
129     10465
130     10467
131     10039
132     10037
133     11102
134     10174
135     11371
136     11238
137     10004
138     11105
139     10170
140     10002
141     10009
142     11221
143     11375
144     11232
145     11203
146     10003
147     10803
148     10115
149     10112
150     11365
151     10970
152     10459
153     11432
154     10460
155     11358
156     11421
157     11418
158     11021
159     11355
160     11234
161     11417
162     11228
163     10471
164     11694
165     10470
166     10704
167     11020
168     10474
169     11204
170     10466
171     11436
172     11223
173     10475
174     10701
175     10271
176     11363
177     10305
178     11219
179     10314
180     11229
181     10306
182     11795
183     11366
184     11224
185     11364
186     10543
187     11412
188     11413
189     10310
190     11360
191     11214
192     11423
193     11570
194     10708
195     11357
196     10304
197     10707
198     11001
199     11030
200     10801
201     11426
202     10464
203     10302
204     11783
205     10303
206     10705
207     11427
208     11003
209     11563
210     11422
211     11429
212     11428
213     11239
214     11590
215     10805
216     10552
217     11411
218     10550
219     11414
220     11801
221     11580
222     11581
223     12771
224     11735
225     10606
226     10583
227     10549
228     10301
229     11561
230     11024
231     11693
232     11550
233     11691
234     11803
235     10601
236     11553
237     11552
238     11530
239     11425
240     12569
241     10804
242     10706
243     11692
244     10570
245     10607
246     10591
247     11560
248     12052
249     11793
250     11756
251     11717
252     10709
253     11577
254     11743
255     10977
256     11004
257     11010
258     10901
259     11501
260     11576
261     10605
262     10502
263     10603
264     11779
265     10309
266     10962
267     11797
268     10595
269     10580
270     11762
271     11710
272     11520
273     11362
274     11747
275     11753
276     10703
277     11554
278     11568
279     11596
280     11096
281     11746
282     10504
283     10965
284     11791
2

In [21]:
zip_unique['lat'] = zip_unique.progress_apply(postcode_to_lat, axis=1)
zip_unique['lon'] = zip_unique.progress_apply(postcode_to_lon, axis=1)

100%|████████████████████████████████████████████████████████████████████████████████| 374/374 [03:06<00:00,  2.00it/s]


In [56]:
zip_unique

zip_code        lat        lon
0       11106  40.761229 -73.929593
1       10006  40.708312 -74.013440
2       11201  40.692500 -73.991763
3       11430  40.653199 -73.784133
4       10032  40.837391 -73.941015
5       10075  40.773597 -73.956517
6       10069  40.776940 -73.988253
7       10001  40.748426 -73.994117
8       10019  40.763992 -73.985893
9       11216  40.680521 -73.949349
10      10169  40.754515 -73.975835
11      10455  40.815454 -73.908826
12      10025  40.796283 -73.969972
13      11218  40.643656 -73.977282
14      10016  40.745486 -73.979671
15      11205  40.693085 -73.964395
16      10162  40.769699 -73.950719
17      11103  40.762559 -73.913691
18      10022  40.758279 -73.967885
19      10128  40.781741 -73.951147
20      10020  40.758980 -73.979655
21      10010  40.739885 -73.985222
22      11104  40.744043 -73.921136
23      10030  40.818065 -73.943105
24      11207  40.672345 -73.894855
25      11419  40.688206 -73.823418
26      10452  40.837781 -73.922321
27      10472  40.829796 -73.868743
28      10279  40.712369 -74.008401
29      11208  40.676360 -73.874577
30      11420  40.675374 -73.815064
31      11361  40.763342 -73.774271
32      11434  40.677314 -73.779485
33      11212  40.662012 -73.914759
34      10031  40.825310 -73.948360
35      11206  40.701645 -73.943105
36      10111  40.759220 -73.977836
37      10005  40.720452 -74.006706
38      10103  40.760277 -73.976285
39      10028  40.776539 -73.953739
40      10278  40.715485 -74.004193
41      11385  40.703925 -73.892403
42      10012  40.725063 -73.998065
43      11416  40.684466 -73.850384
44      11356  40.785984 -73.844770
45      10282  40.715924 -74.015087
46      11435  40.699612 -73.808415
47      11424  40.715700 -73.826700
48      10461  40.845859 -73.841064
49      11235  40.586189 -73.951600
50      11236  40.640007 -73.902992
51      10458  40.861559 -73.888772
52      11373  40.739766 -73.878525
53      10065  40.766073 -73.964662
54      10014  40.734090 -74.004893
55      10024  40.786271 -73.975631
56      11237  40.701298 -73.918203
57      11109  40.746276 -73.956841
58      10110  40.753818 -73.981097
59      11231  40.680265 -74.000159
60      11377  40.744809 -73.904360
61      11213  40.670881 -73.937523
62      11211  40.725539 -73.946395
63      10168  40.751172 -73.976660
64      10034  40.865760 -73.921664
65      11367  40.728792 -73.821065
66      10011  40.740868 -73.999447
67      10153  40.763541 -73.972212
68      10171  40.756202 -73.974574
69      10454  40.807741 -73.918223
70      10119  40.750530 -73.993090
71      10165  40.752303 -73.978431
72      10167  40.754649 -73.974771
73      11233  40.678977 -73.920300
74      11225  40.662458 -73.952910
75      10029  40.792048 -73.944442
76      11217  40.682279 -73.979094
77      11514  40.750407 -73.612034
78      10036  40.759498 -73.989216
79      11368  40.748088 -73.860623
80      11370  40.762389 -73.890554
81      11372  40.751383 -73.883173
82      10468  40.865509 -73.900245
83      10154  40.757866 -73.972311
84      11374  40.724962 -73.861919
85      10453  40.852335 -73.911997
86      11378  40.724009 -73.900127
87      11415  40.706730 -73.830780
88      10462  40.842893 -73.857949
89      11230  40.621524 -73.965614
90      11210  40.627427 -73.945651
91      11433  40.697250 -73.786894
92      11101  40.748484 -73.938832
93      10035  40.796804 -73.939937
94      10280  40.707994 -74.017197
95      11749  40.806032 -73.171600
96      10023  40.776973 -73.981065
97      10033  40.849317 -73.935206
98      10013  40.723075 -74.009359
99      10017  40.750998 -73.993697
100     10463  40.880988 -73.905670
101     10007  40.714050 -74.008219
102     11788  40.820629 -73.202096
103     10173  40.753889 -73.979539
104     10456  40.829743 -73.908570
105     10457  40.847962 -73.897723
106     10040  40.852883 -73.932484
107     11369  40.762495 -73.872836
108     10044  40.761924 -73.949943
109     10473  40.817776 -73.856840
110     1

In [53]:
zip_unique.loc[zip_unique['lat'] == 0]

Empty DataFrame
Columns: [zip_code, lat, lon]
Index: []

In [36]:
zip_unique['lat'] = pd.to_numeric(zip_unique['lat'])
zip_unique['lon'] = pd.to_numeric(zip_unique['lon'])

> API를 통해 나오지 않는 정보는 직접 검색해 찾아봅니다.
https://www.freemaptools.com/convert-us-zip-code-to-lat-lng.htm

In [54]:
zip_unique.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374 entries, 0 to 373
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   zip_code  374 non-null    int64  
 1   lat       374 non-null    float64
 2   lon       374 non-null    float64
dtypes: float64(2), int64(1)
memory usage: 8.9 KB


In [55]:
zip_unique.loc[zip_unique['zip_code']==11424, 'lat'] = 40.71570
zip_unique.loc[zip_unique['zip_code']==11424, 'lon'] = -73.82670

zip_unique.loc[zip_unique['zip_code']==10119, 'lat'] = 40.75053
zip_unique.loc[zip_unique['zip_code']==10119, 'lon'] = -73.99309

zip_unique.loc[zip_unique['zip_code']==11351, 'lat'] = 40.75530
zip_unique.loc[zip_unique['zip_code']==11351, 'lon'] = -73.82680

In [69]:
zip_unique

zip_code        lat        lon
0       11106  40.761229 -73.929593
1       10006  40.708312 -74.013440
2       11201  40.692500 -73.991763
3       11430  40.653199 -73.784133
4       10032  40.837391 -73.941015
5       10075  40.773597 -73.956517
6       10069  40.776940 -73.988253
7       10001  40.748426 -73.994117
8       10019  40.763992 -73.985893
9       11216  40.680521 -73.949349
10      10169  40.754515 -73.975835
11      10455  40.815454 -73.908826
12      10025  40.796283 -73.969972
13      11218  40.643656 -73.977282
14      10016  40.745486 -73.979671
15      11205  40.693085 -73.964395
16      10162  40.769699 -73.950719
17      11103  40.762559 -73.913691
18      10022  40.758279 -73.967885
19      10128  40.781741 -73.951147
20      10020  40.758980 -73.979655
21      10010  40.739885 -73.985222
22      11104  40.744043 -73.921136
23      10030  40.818065 -73.943105
24      11207  40.672345 -73.894855
25      11419  40.688206 -73.823418
26      10452  40.837781 -73.922321
27      10472  40.829796 -73.868743
28      10279  40.712369 -74.008401
29      11208  40.676360 -73.874577
30      11420  40.675374 -73.815064
31      11361  40.763342 -73.774271
32      11434  40.677314 -73.779485
33      11212  40.662012 -73.914759
34      10031  40.825310 -73.948360
35      11206  40.701645 -73.943105
36      10111  40.759220 -73.977836
37      10005  40.720452 -74.006706
38      10103  40.760277 -73.976285
39      10028  40.776539 -73.953739
40      10278  40.715485 -74.004193
41      11385  40.703925 -73.892403
42      10012  40.725063 -73.998065
43      11416  40.684466 -73.850384
44      11356  40.785984 -73.844770
45      10282  40.715924 -74.015087
46      11435  40.699612 -73.808415
47      11424  40.715700 -73.826700
48      10461  40.845859 -73.841064
49      11235  40.586189 -73.951600
50      11236  40.640007 -73.902992
51      10458  40.861559 -73.888772
52      11373  40.739766 -73.878525
53      10065  40.766073 -73.964662
54      10014  40.734090 -74.004893
55      10024  40.786271 -73.975631
56      11237  40.701298 -73.918203
57      11109  40.746276 -73.956841
58      10110  40.753818 -73.981097
59      11231  40.680265 -74.000159
60      11377  40.744809 -73.904360
61      11213  40.670881 -73.937523
62      11211  40.725539 -73.946395
63      10168  40.751172 -73.976660
64      10034  40.865760 -73.921664
65      11367  40.728792 -73.821065
66      10011  40.740868 -73.999447
67      10153  40.763541 -73.972212
68      10171  40.756202 -73.974574
69      10454  40.807741 -73.918223
70      10119  40.750530 -73.993090
71      10165  40.752303 -73.978431
72      10167  40.754649 -73.974771
73      11233  40.678977 -73.920300
74      11225  40.662458 -73.952910
75      10029  40.792048 -73.944442
76      11217  40.682279 -73.979094
77      11514  40.750407 -73.612034
78      10036  40.759498 -73.989216
79      11368  40.748088 -73.860623
80      11370  40.762389 -73.890554
81      11372  40.751383 -73.883173
82      10468  40.865509 -73.900245
83      10154  40.757866 -73.972311
84      11374  40.724962 -73.861919
85      10453  40.852335 -73.911997
86      11378  40.724009 -73.900127
87      11415  40.706730 -73.830780
88      10462  40.842893 -73.857949
89      11230  40.621524 -73.965614
90      11210  40.627427 -73.945651
91      11433  40.697250 -73.786894
92      11101  40.748484 -73.938832
93      10035  40.796804 -73.939937
94      10280  40.707994 -74.017197
95      11749  40.806032 -73.171600
96      10023  40.776973 -73.981065
97      10033  40.849317 -73.935206
98      10013  40.723075 -74.009359
99      10017  40.750998 -73.993697
100     10463  40.880988 -73.905670
101     10007  40.714050 -74.008219
102     11788  40.820629 -73.202096
103     10173  40.753889 -73.979539
104     10456  40.829743 -73.908570
105     10457  40.847962 -73.897723
106     10040  40.852883 -73.932484
107     11369  40.762495 -73.872836
108     10044  40.761924 -73.949943
109     10473  40.817776 -73.856840
110     1

## 💻 기존 데이터셋과 합치기

> 기존 df의 순서를 지키며 merge

In [70]:
df = df.merge(df.merge(zip_unique, how='left', sort=False))

In [71]:
df

zip_code          pickup_hour  month  day  weekday  hour  is_weekend  \
0         11106  2015-01-01 00:00:00      1    1        3     0           0   
1         10006  2015-01-01 00:00:00      1    1        3     0           0   
2         11201  2015-01-01 00:00:00      1    1        3     0           0   
3         11430  2015-01-01 00:00:00      1    1        3     0           0   
4         10032  2015-01-01 00:00:00      1    1        3     0           0   
...         ...                  ...    ...  ...      ...   ...         ...   
87015     11219  2015-01-31 23:00:00      1   31        5    23           1   
87016     10456  2015-01-31 23:00:00      1   31        5    23           1   
87017     11355  2015-01-31 23:00:00      1   31        5    23           1   
87018     11210  2015-01-31 23:00:00      1   31        5    23           1   
87019     10167  2015-01-31 23:00:00      1   31        5    23           1   

       cnt        lat        lon  
0       75  40.761229 -73.929593  
1       64  40.708312 -74.013440  
2      169  40.692500 -73.991763  
3      222  40.653199 -73.784133  
4       64  40.837391 -73.941015  
...    ...        ...        ...  
87015    1  40.632650 -73.996601  
87016    3  40.829743 -73.908570  
87017    1  40.750423 -73.819936  
87018    1  40.627427 -73.945651  
87019    1  40.754649 -73.974771  

[87020 rows x 10 columns]

## 💻 Folium을 활용한 Heatmap 만들기

In [89]:
def generateBaseMap(default_location=[40.746176, -73.996453], # Boston의 좌표
                    default_zoom_start=9): #location과 시작 확대 정도를 파라미터로 받는 함수를 만들었어요
    base_map = folium.Map(location=default_location, control_scale=True, zoom_start=default_zoom_start)
    return base_map

In [90]:
base_map=generateBaseMap()
base_map

In [91]:
from folium.plugins import HeatMap
HeatMap(data=df[['lat', 'lon']].drop_duplicates().reset_index(drop=True), 
        radius=8, max_zoom=5).add_to(base_map)

In [92]:
base_map

## 💻 추가된 정보로 csv 파일 생성

In [94]:
df.to_csv('NYC_taxi_lat_lon.csv', index=False)